dev-tools/smoke/smoke_common.py

```
@@ -171,6 +175,10 @@ def run_transport_session(
         try:
             with closing(open_session()) as session:
                 return execute_session(session)
+        except TransientUpstreamSmokeFailure:
+            if deadline is None or time.time() >= deadline:
+                raise
+            time.sleep(retry_interval_seconds)
```

Ruff (0.15.21)

[warning] `try`-`except` within a loop incurs performance overhead

(PERF203)